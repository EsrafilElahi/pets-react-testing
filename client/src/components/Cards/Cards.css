.pet-cards-container {
  width: 80%;
  display: flex;
  flex-wrap: wrap;
}
