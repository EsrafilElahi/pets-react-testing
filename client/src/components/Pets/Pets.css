.app-container {
  display: flex;
  padding: 3rem 0;
}
