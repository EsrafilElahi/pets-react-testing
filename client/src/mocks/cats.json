[
  {
    "id": 1,
    "name": "Sydney",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1529778873920-4da4926a72c2?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MTJ8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "female"
  },
  {
    "id": 2,
    "name": "Ja",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1536590158209-e9d615d525e4?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MTl8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "male"
  },
  {
    "id": 3,
    "name": "Holly",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1606214174585-fe31582dc6ee?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MjB8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "female"
  },
  {
    "id": 4,
    "name": "Kevin",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1592194996308-7b43878e84a6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8OXx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "male"
  },
  {
    "id": 5,
    "name": "Rachel",
    "phone": "333-333-3333",
    "email": "laith@hotmail.com",
    "image": {
      "url": "https://images.unsplash.com/photo-1561948955-570b270e7c36?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8NXx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=800&q=60",
      "alt": "beautiful cat"
    },
    "favoured": false,
    "color": "grey",
    "gender": "female"
  }
]
