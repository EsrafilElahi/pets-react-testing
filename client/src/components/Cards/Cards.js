import { useContext } from "react";
import Card from "../Card/Card";
import { PetsContext } from "../Pets/Pets";
import "./Cards.css";

const Cards = () => {
  const { cats } = useContext(PetsContext);

  return (
    <div className="pet-cards-container">
      {cats.map((cat, index) => {
        return (
          <Card
            key={cat.id}
            name={cat.name}
            phone={cat.phone}
            email={cat.email}
            image={cat.image}
            favoured={cat.favoured}
            index={index}
          />
        );
      })}
    </div>
  );
};

export default Cards;
