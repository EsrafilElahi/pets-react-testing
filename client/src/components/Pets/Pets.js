import "./Pets.css";
import Filter from "../Filter/Filter";
import Cards from "../Cards/Cards";

import axios from "axios";
import { useState, useEffect, createContext } from "react";

export const PetsContext = createContext({
  cats: [],
  setCats: () => {},
});

const Pets = () => {
  const [cats, setCats] = useState([]);
  const [filteredCats, setFilteredCats] = useState([]);
  const [filters, setFilters] = useState({
    gender: "any",
    favoured: "any",
  });

  const fetchCats = async () => {
    const response = await axios.get("http://localhost:4000/cats");
    setCats(response.data);
    setFilteredCats(response.data);
  };

  useEffect(() => {
    fetchCats();
  }, []);

  useEffect(() => {
    let catsFiltered = [...cats];
    if (filters.gender !== "any") {
      catsFiltered = catsFiltered.filter(
        (cat) => cat.gender === filters.gender
      );
    }
    if (filters.favoured !== "any") {
      catsFiltered = catsFiltered.filter((cat) => {
        console.log(
          cat.favoured,
          filters.favoured,
          cat.favoured === filters.favoured
        );
        return (
          cat.favoured === (filters.favoured === "favoured" ? true : false)
        );
      });
    }
    setFilteredCats(catsFiltered);
  }, [filters]);

  return (
    <div className="container">
      <div className="app-container">
        <PetsContext.Provider value={{ cats: filteredCats, setCats }}>
          <Filter filters={filters} setFilters={setFilters} />
          <Cards />
        </PetsContext.Provider>
      </div>
    </div>
  );
};

export default Pets;
