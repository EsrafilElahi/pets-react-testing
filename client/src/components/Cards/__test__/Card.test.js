import { screen, render } from '@testing-library/react';
import Cards from '../Cards';
import cats from '../../../mocks/cats.json';
import { PetsContext } from '../../Pets/Pets';


describe("Cards", () => {
  test("render 5 cards in page", () => {
    // render PetsContext with values
    render(
      <PetsContext.Provider
        value={{
          cats: cats,
          setCats: () => { },
        }}
      >
        <Cards />
      </PetsContext.Provider>
    );
    // assertion
    const cards = screen.getAllByRole("article");
    expect(cards).toHaveLength(5)
  });

  test("snapshot test", () => {
    const { asFragment } = render(
      <PetsContext.Provider
        value={{
          cats: cats,
          setCats: () => { },
        }}
      >
        <Cards />
      </PetsContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot()
  })
});