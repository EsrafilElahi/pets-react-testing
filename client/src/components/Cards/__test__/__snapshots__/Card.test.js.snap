// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards snapshot test 1`] = `
<DocumentFragment>
  <div
    class="pet-cards-container"
  >
    <article
      class="card"
    >
      <div
        class="card-header"
      >
        <img
          alt="beautiful cat"
          class="card-img"
          src="https://images.unsplash.com/photo-1529778873920-4da4926a72c2?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MTJ8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60"
        />
        <button
          class="heart"
        >
          <img
            alt="outlined heart"
            src="heartOutlined.svg"
          />
        </button>
      </div>
      <div
        class="card-content"
      >
        <h3>
          One
        </h3>
        <p>
          333-333-3333
        </p>
        <p>
          laith@hotmail.com
        </p>
        <p>
          female
        </p>
      </div>
    </article>
    <article
      class="card"
    >
      <div
        class="card-header"
      >
        <img
          alt="beautiful cat"
          class="card-img"
          src="https://images.unsplash.com/photo-1536590158209-e9d615d525e4?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MTl8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60"
        />
        <button
          class="heart"
        >
          <img
            alt="outlined heart"
            src="heartOutlined.svg"
          />
        </button>
      </div>
      <div
        class="card-content"
      >
        <h3>
          Two
        </h3>
        <p>
          333-333-3333
        </p>
        <p>
          laith@hotmail.com
        </p>
        <p>
          male
        </p>
      </div>
    </article>
    <article
      class="card"
    >
      <div
        class="card-header"
      >
        <img
          alt="beautiful cat"
          class="card-img"
          src="https://images.unsplash.com/photo-1606214174585-fe31582dc6ee?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MjB8fGNhdHxlbnwwfHwwfHw%3D&auto=format&fit=crop&w=800&q=60"
        />
        <button
          class="heart"
        >
          <img
            alt="outlined heart"
            src="heartOutlined.svg"
          />
        </button>
      </div>
      <div
        class="card-content"
      >
        <h3>
          Three
        </h3>
        <p>
          333-333-3333
        </p>
        <p>
          laith@hotmail.com
        </p>
        <p>
          female
        </p>
      </div>
    </article>
    <article
      class="card"
    >
      <div
        class="card-header"
      >
        <img
          alt="beautiful cat"
          class="card-img"
          src="https://images.unsplash.com/photo-1592194996308-7b43878e84a6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8OXx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=800&q=60"
        />
        <button
          class="heart"
        >
          <img
            alt="outlined heart"
            src="heartOutlined.svg"
          />
        </button>
      </div>
      <div
        class="card-content"
      >
        <h3>
          Four
        </h3>
        <p>
          333-333-3333
        </p>
        <p>
          laith@hotmail.com
        </p>
        <p>
          male
        </p>
      </div>
    </article>
    <article
      class="card"
    >
      <div
        class="card-header"
      >
        <img
          alt="beautiful cat"
          class="card-img"
          src="https://images.unsplash.com/photo-1561948955-570b270e7c36?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8NXx8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=800&q=60"
        />
        <button
          class="heart"
        >
          <img
            alt="outlined heart"
            src="heartOutlined.svg"
          />
        </button>
      </div>
      <div
        class="card-content"
      >
        <h3>
          Five
        </h3>
        <p>
          333-333-3333
        </p>
        <p>
          laith@hotmail.com
        </p>
        <p>
          female
        </p>
      </div>
    </article>
  </div>
</DocumentFragment>
`;
